### DISCUSSION TITLE
Some Questions about Brusselator Example

### DISCUSSION THREAD

**Discussion by yangheart** [zwicker-group/py-pde#96] [Q&A]
Dear Dr. Zwicker,

I am really appreciated that you developed such an amazing package for us to solve PDE. I am really interested in the example, called Brusselator - Using the PDE class. But there are conditions there which I am still confused about. Could you help me to figure out them?

Firstly, does t_range=20 mean t $\in$ [0,20] or mean t$\in$ [0, 20dt]?

Secondly, for bc="natural", does it mean there is no bounday condition?

Thirdly, If I set up bc="neumann", does it means neumann condition is 0?

Thank you so much!

**Comment by david-zwicker** [comment]
Great to hear that you like our work! To your questions:

1. The former: `t_range` specifies the time range in time units not in the number of time steps
2. Natural boundary conditions are 'periodic' for periodic grid axes and 'neumann' for non-periodic grid axes
3. Indeed, 'neumann' implies a vanishing derivative at the boundary

You can read more about [boundary conditions in the documentation](https://py-pde.readthedocs.io/en/latest/packages/pde.grids.boundaries.html).

**Reply by yangheart** [reply]
Dear Dr. Zwicker,

Thank you so much! I have one more question. How could I save the data for every time step?

**Reply by david-zwicker** [reply]
You can simply set an interval for the tracker that is smaller than your timestep `dt`. When doing this, you might want to use the latest master branch of the package, since we fixed a bug that might slow down your simulation in this case.

**Reply by yangheart** [reply]
Dear Dr. Zwicker,

Sorry for disturbing again. I am still confused about the conditions of this example. From the paper, it said that: "The standard PDE classes support additive Gaussian white noise, but alternatives, including multiplicative noise, can be specified in userdefined classes." Does it mean that the main body of the PDE in the Brusselator Example involves a Gaussian noise term?

Thank you so much!

**Reply by david-zwicker** [reply]
Which version of the Brusselator Example do you refer to? Could you give a link to the example? Generally, the `py-pde` package support stochastic partial differential equations, but the actual implementations vary a bit. Also, the paper does not refer to the newest version of the package, so some of its descriptions might be outdated.

**Reply by yangheart** [reply]
Here is the example of Brusselator that I refer to:
https://py-pde.readthedocs.io/en/latest/examples_gallery/pde_brusselator_expression.html#sphx-glr-examples-gallery-pde-brusselator-expression-py

The equation on this page seems to be deterministic since the white noises are only added to the initial values. I am wondering if any stochastic terms are involved in the main body of the equation besides the initial values?

**Reply by david-zwicker** [reply]
This example solves the deterministic equations for the Brusselator, starting from random initial conditions, as you suspected. However, if you add the `noise` argument to the call to `PDE`, you wind up with the stochastic version, as described in the [documentation of `PDE`](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.html#pde.pdes.pde.PDE). In this case uncorrelated Gaussian white noise is added to both fields and the value supplied by the `noise` argument sets the noise strength.
